src/convert.rs

```diff
@@ -1,69 +1,56 @@
 //! Conversion between bls12_381 types and bytes.
 
-use crate::error::{Error, Result};
-use group::{CurveAffine, CurveProjective, EncodedPoint};
+use blst::{blst_fr, blst_fr_from_scalar, blst_scalar};
 
-use crate::{PK_SIZE, SIG_SIZE, SK_SIZE};
-use ff::PrimeField;
-use pairing::bls12_381::{Fr, FrRepr, G1Affine, G2Affine, G1, G2};
+use crate::{
+    error::{Error, Result},
+    Fr, DST, G1, G2, PK_SIZE, SIG_SIZE, SK_SIZE,
+};
 
-/// Convert big endian bytes to bls12_381::Fr
-pub fn fr_from_be_bytes(bytes: [u8; SK_SIZE]) -> Result<Fr> {
-    let mut le_bytes = bytes;
-    le_bytes.reverse();
-    let mut fr_u64s = [0u64; 4];
-    for i in 0..4 {
-        let mut next_u64_bytes = [0u8; 8];
-        for j in 0..8 {
-            next_u64_bytes[j] = le_bytes[i * 8 + j];
-        }
-        fr_u64s[i] = u64::from_le_bytes(next_u64_bytes);
-    }
-    Ok(Fr::from_repr(FrRepr(fr_u64s))?)
-}
-
-/// Convert bls12_381::Fr to big endian bytes
-pub fn fr_to_be_bytes(fr: Fr) -> [u8; SK_SIZE] {
-    let mut bytes = [0u8; SK_SIZE];
-    // iterating 4 u64s which are in order suiting little endian bytes
-    // and must be reversed to get big endian bytes
-    fr.into_repr().0.iter().enumerate().for_each(|(le_i, n)| {
-        let be_i = 4 - le_i - 1;
-        n.to_be_bytes().iter().enumerate().for_each(|(j, byte)| {
-            bytes[8 * be_i + j] = *byte;
-        });
-    });
-    bytes
+pub(crate) fn derivation_index_into_fr(index: &[u8]) -> Fr {
+    hash_to_field(index)
 }
 
-/// Convert big endian bytes to bls12_381::G1
-pub fn g1_from_be_bytes(bytes: [u8; PK_SIZE]) -> Result<G1> {
-    let mut compressed: <G1Affine as CurveAffine>::Compressed = EncodedPoint::empty();
-    compressed.as_mut().copy_from_slice(&bytes);
-    let opt_affine = compressed.into_affine().ok();
-    let projective = opt_affine.ok_or(Error::InvalidBytes)?.into_projective();
-    Ok(projective)
+/// Generates a scalar as described in IETF hash to curve
+/// https://www.ietf.org/archive/id/draft-irtf-cfrg-hash-to-curve-10.html#name-hashing-to-a-finite-field-2
+/// There are two main candidates for converting arbitrary bytes to scalar:
+/// * hash to field as described in hash to curve
+/// * bls signature keygen based on hkdf
+/// BLS Signature spec has strong recommendations about IKM which don't apply when deriving indexes.
+/// Hash to field is also slightly faster than HKDF.
+fn hash_to_field<T: AsRef<[u8]>>(msg: T) -> Fr {
+    // TODO IC remove unwrap here
+    let scalar = blst_scalar::hash_to(msg.as_ref(), DST).unwrap();
+    let mut fr = blst_fr::default();
+    unsafe {
```

Comment:
unsafe.... no way to do this using blstrs?

Reply:
Ah yes, there's a try_into on [scalar.rs#L218](https://github.com/filecoin-project/blstrs/blob/07a84f9727d2fb7387efb2df0f17ddc5d35571d3/src/scalar.rs#L218), I'll use that. Good catch, I'll update this PR with this change.

Reply:
TryInto is now being used instead of unsafe. See [convert.rs#L31-L34](https://github.com/maidsafe/blsttc/pull/42/commits/3956a2c25a65ab1b9d7bcdadc09450278430c046#diff-d42cae02d6ab6561d7a3d9a4d5538b03cdb33e075dfe41a89e70fd8206d9eaabR31-R34)